subsys/bluetooth/services/dis.c

```diff
@@ -230,6 +351,159 @@ static int dis_set(const char *name, size_t len_rd,
 		}
 		return 0;
 	}
+#endif
+#if defined(CONFIG_BT_DIS_UDI)
+	const size_t currently_used_udi_merged_space = 1 +
+		strlen(BT_DIS_UDI_LABEL_STR_REF) +
+		strlen(BT_DIS_UDI_DI_STR_REF) +
+		strlen(BT_DIS_UDI_ISSUER_STR_REF) +
+		strlen(BT_DIS_UDI_AUTHORITY_STR_REF) +
+		BT_DIS_UDI_FLAG_LABEL +
+		BT_DIS_UDI_FLAG_DI +
+		BT_DIS_UDI_FLAG_ISSUER +
+		BT_DIS_UDI_FLAG_AUTHORITY;
+
+	if (!strncmp(name, "udi_label", nlen)) {
+		size_t in_use_without_this = currently_used_udi_merged_space
+			- strlen(BT_DIS_UDI_LABEL_STR_REF) - BT_DIS_UDI_FLAG_LABEL;
+		if (sizeof(bt_dis_udi_merged) < in_use_without_this + strlen(store) + 1)
+		{
+			LOG_ERR("Failed to set UDI Label. Not enough space. The sum of the 4 DIS UDI for "
+				"Medical Devices strings may not exceed the maximum attribute length"
+				" or maximum transmit MTU.");
+			return 0;
+		}
+
+		len = read_cb(store, &dis_udi_label, sizeof(dis_udi_label) - 1);
+		if (len < 0) {
+			LOG_ERR("Failed to read UDI Label from storage"
+				       " (err %zd)", len);
+		} else {
+			dis_udi_label[len] = '\0';
+			bt_dis_udi_merged_size = 0;
+
+			LOG_DBG("UDI Label set to %s", dis_udi_label);
+		}
+		return 0;
+	}
+
+	if (!strncmp(name, "udi_di", nlen)) {
+		size_t in_use_without_this = currently_used_udi_merged_space
+			- strlen(BT_DIS_UDI_DI_STR_REF) - BT_DIS_UDI_FLAG_DI;
+		if (sizeof(bt_dis_udi_merged) < in_use_without_this + strlen(store) + 1)
+		{
+			LOG_ERR("Failed to set UDI Label. Not enough space. The sum of the 4 DIS UDI for "
+				"Medical Devices strings may not exceed the maximum attribute length"
+				" or maximum transmit MTU.");
+			return 0;
+		}
+
+		len = read_cb(store, &dis_udi_di, sizeof(dis_udi_di) - 1);
+		if (len < 0) {
+			LOG_ERR("Failed to read UDI Device Identifier from storage"
+				       " (err %zd)", len);
+		} else {
+			dis_udi_di[len] = '\0';
+			bt_dis_udi_merged_size = 0;
+
+			LOG_DBG("UDI Device Identifier set to %s", dis_udi_di);
+		}
+		return 0;
+	}
+
+	if (!strncmp(name, "udi_issuer", nlen)) {
+		size_t in_use_without_this = currently_used_udi_merged_space
+			- strlen(BT_DIS_UDI_ISSUER_STR_REF) - BT_DIS_UDI_FLAG_ISSUER;
+		if (sizeof(bt_dis_udi_merged) < in_use_without_this + strlen(store) + 1)
+		{
+			LOG_ERR("Failed to set UDI Label. Not enough space. The sum of the 4 DIS UDI for "
+				"Medical Devices strings may not exceed the maximum attribute length"
+				" or maximum transmit MTU.");
```

Comment:
> may not exceed the [...] maximum transmit MTU

Is this mandated by spec? GATT has a 'long read' procedure that's used to read longer values than ATT_MTU. We can just make sure long reads are working.

Reply:
This is not in the spec. I have removed it.